master/buildbot/data/buildrequests.py

```diff
@@ -62,6 +64,39 @@ def startConsuming(self, callback, options, kwargs):
         return self.master.mq.startConsuming(callback,
                                              ('buildrequests', buildrequestid, None))
 
+    @defer.inlineCallbacks
+    def control(self, action, args, kwargs):
+        if action != "cancel":
+            raise ValueError("action: {} is not supported".format(action))
+        brid = kwargs['buildrequestid']
+        # first, try to claim the request; if this fails, then it's too late to
+        # cancel the build anyway
+        try:
+            b = yield self.master.db.buildrequests.claimBuildRequests(brids=[brid])
+        except AlreadyClaimedError:
+            # XXX race condition
+            # - After a buildrequest was claimed, and
+            # - Before creating a build,
+            # the claiming master still
+            # needs to do some processing, (send a message to the message queue,
+            # call maybeStartBuild on the related builder).
+            # In that case we won't have the related builds here. We don't have
+            # an alternative to letting them run without stopping them for now.
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
+
+            # Don't call the data API here, as the buildrequests might have been
+            # taken by another master. We just send the stop message and forget about those.
```

Comment:
It might be faster, if you set your "self.master.mq.produce" kwargs dict before looping, avoiding to looking up into kwargs at each iteration